spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala

```diff
@@ -115,8 +117,57 @@ class DeltaHistoryManager(
           }
         }.map(DeltaHistory.fromCommitInfo)
       }
-    // Spark should return the commits in increasing order as well
-    monotonizeCommitTimestamps(info.collect()).reverse
+    val monotonizedCommits = if (isRangeICT) {
+      info.collect()
+    } else {
+      monotonizeCommitTimestamps(info.collect())
+    }
+    monotonizedCommits.reverse
+  }
+
+  /**
+   * Get the commit information of the Delta table from commit `[start, end]`. If `end` is `None`,
+   * we return all commits from start to now.
+   * @param start The start of the commit range, inclusive.
+   * @param end The end of the commit range, inclusive.
+   */
+  def getHistory(
+      start: Long,
+      end: Option[Long] = None): Seq[DeltaHistory] = {
+    val currentSnapshot = deltaLog.unsafeVolatileSnapshot
+    val (snapshot, upperBoundExclusive) = end match {
+        case Some(endInclusive) if currentSnapshot.version >= endInclusive =>
+          // Use the cache snapshot if it's fresh enough for the [start, endInclusive] query.
+          (currentSnapshot, math.min(currentSnapshot.version, endInclusive) + 1)
+        case _ =>
+          // Either end doesn't exist or the currently cached snapshot isn't new enough to
+          // satisfy it.
+          val snapshot = deltaLog.update()
+          val endInclusive = end.getOrElse(snapshot.version).min(snapshot.version)
+          (snapshot, endInclusive + 1)
+      }
+    if (!DeltaConfigs.IN_COMMIT_TIMESTAMPS_ENABLED.fromMetaData(snapshot.metadata)) {
+      getHistoryImpl(start, upperBoundExclusive, isRangeICT = false, snapshot)
+    } else {
+      InCommitTimestampUtils.getValidatedICTEnablementInfo(snapshot.metadata) match {
+        case Some(Commit(ictEarliest, _)) =>
+          val nonICTCommits = if (ictEarliest > start) {
+            getHistoryImpl(
+              start, math.min(upperBoundExclusive, ictEarliest), isRangeICT = false, snapshot)
+          } else {
+            Seq.empty
+          }
+          val ictCommits = if (upperBoundExclusive > ictEarliest) {
+            getHistoryImpl(
+              math.max(ictEarliest, start), upperBoundExclusive, isRangeICT = true, snapshot)
+          } else {
+            Seq.empty
+          }
+          ictCommits ++ nonICTCommits
```

Comment:
The ordering here is important but it requires the reader to reason about the ordering of reverse and appending lists here to prove correctness. I'm not sure if there's a more intuitive way to order them here though.

We can add some comments:
```
// Merge the two sequences, ensuring ICT commits are listed first as they are more recent,
// followed by non-ICT commits, maintaining the reverse chronological order
```